docs/rfcs/030-vectored-timeline-get.md

```diff
@@ -0,0 +1,78 @@
+# Vectored Timeline Get
+
+Created on: 2024-01-02
+Author: Christian Schwarz
+
+# Summary
+
+A brief RFC / GitHub Epic describing a vectored version of the `Timeline::get` method that is at the heart of Pageserver.
+
+# Motivation
+
+During basebackup, we issue many `Timeline::get` calls for SLRU pages that are *adjacent* in key space.
+For an example, see
+https://github.com/neondatabase/neon/blob/5c88213eaf1b1e29c610a078d0b380f69ed49a7e/pageserver/src/basebackup.rs#L281-L302.
+
+Each call must traverse the layer map to gather reconstruct data (`Timeline::get_reconstruct_data`) for the requested page number (`blknum` in the example).
+
+That is quite inefficient because:
+
+1. We do the layer map traversal repeatedly, even if, e.g., all the data sits in the same image layer at the bottom of the stack.
+2. Anecdotally, keys adjacent in keyspace and written simultaneously also end up physically adjacent in the layer files [^1].
+   So, to provide the reconstruct data for N adjacent keys, we would actually only _need_ to issue a single large read to the filesystem, instead of the N reads we currently do.
+   The filesystem, in turn, ideally stores the layer file physically contiguously, so our large read will turn into one IOP toward the disk.
```

Comment:
The impact on disk I/O depends on the size of each SLRU block we're reading -- many modern NVMe drives can saturate bandwidth at the size of our ~8kiB blocks, so aggregating these into larger system calls isn't a big efficiency win.

I think it's reasonable to expect that delta layer entries for SLRU blocks will typically be much smaller than 8kiB (therefore this optimization is helpful), but it would be interesting to check that theory while working on this.

Reply:
At least in my investigation it wasn't, it read _many_ `256KiB/8KiB` segments.

Reply:
OK - for collections of adjacent block-sized reads, the real benefit we get from coalescing will be that the OS can issue smaller I/Os concurrently.  So we could choose between issuing  large read system calls, or issuing our own concurrent smaller I/Os.  Doing larger system calls (up to some reasonable limit) probably makes sense in this use case where we want the blocks in order anyway